### DISCUSSION TITLE
[LUA] How to get Host input device from ioport input sequences?

### DISCUSSION THREAD

**Discussion by Jakobud** [mamedev/discussions#148] [Q&A] [answered]
I'm looking at the ioports and trying to figure out how to get the Host input device item token for each input.

For example, if I have an input token like `P1_BUTTON1` that is assigned to `KEYCODE_ENTER` how do I get the device item token that says `KEYCODE_ENTER`. I can get the `input_sequence` for an input easily enough:

```
for portname, port in pairs(ioports.ports) do
  for fieldname, field in pairs(port.fields) do
    print(input:seq_name(field:input_seq('standard')))
  end
end
```

So that will print out an input sequence like `Kbd Z`. But I'm looking for the token version of it, like `KEYCODE_Z`. How do I figure out the relationship between an `input_sequence` and the host `input_device`(s) that it uses?

The only way I can think of to accomplish this is iterating through the `input_class`s, then iterating through the `input_device` items and building a hashmap using names and tokens like this

```
deviceHashMap = {
  "Kbd Z" = "KEYCODE_Z",
  "Kbd X" = "KEYCODE_X",
  "Kbd -" = "KEYCODE_MINUS",
  etc...
}
```
Then I would use this hashmap to find the token value from an input sequence.

But I don't know if that would be reliable. And I know Windows does weird things with device classes like detecting multiple HID Keyboards that you would have to iterate through that have item tokens like `KEYBOARD_2_ENTER`. So when would those ever get used? Maybe they wouldn't...  Also I would need to parse the input sequence if there are multiple values like "Kbd ENTER or Kbd CTRL" etc and take different types of sequences into account.

Is there a relationship in the API between input sequences and device input class items?

**Comment by Jakobud** [comment, marked as answer]
Ah I found it. `input:seq_to_tokens(seq)`. I dunno how I missed that. Must have looked right past in the docs like 20 times.

**Comment by cuavas** [comment]
Input sequence objects are considered opaque from the Lua API.
* You can use convert them to or from a token string with the [input manager](https://docs.mamedev.org/luascript/ref-input.html#input-manager) `seq_to_tokens` and `seq_from_tokens` methods.
* You can convert them to a display string with the [input manager](https://docs.mamedev.org/luascript/ref-input.html#input-manager) `seq_name` method.
* You have the user enter one using an [input sequence poller](https://docs.mamedev.org/luascript/ref-input.html#input-sequence-poller).

The included Autofire and Input Macro plugins do various things with input sequences.
